### PR TITLE
[speedup] Use indices mappings instead of deepcopy for all the samples reordering methods

### PR DESCRIPTION
**Pull request by thomwolf**
Use an indices mapping instead of rewriting the dataset for all the samples re-ordering/selection methods (`select`, `sort`, `shuffle`, `shard`, `train_test_split`).

Added a `flatten_indices` method which copy the dataset to a new table to remove the indices mapping with tests.

All the samples re-ordering/selection methods should be a lot faster. The downside is that iterating on very large batch of the dataset might be a little slower when we have changed the order of the samples since with in these case we use `pyarrow.Table.take` instead of `pyarrow.Table.slice`. There is no free lunch but the speed of iterating over the dataset is rarely the bottleneck.

*Backward breaking change*: the `cache_file_name` argument in all the samples re-ordering/selection methods (`select`, `sort`, `shuffle`, `shard`, `train_test_split`) is now called `indices_cache_file_name` on purpose to make it explicit to the user that this caching file is used for caching the indices mapping and not the dataset itself.

### CODE REVIEW

File: src/nlp/arrow_dataset.py [hunk 1]
```diff
@@ -1325,91 +1468,11 @@ def shuffle(
...
             indices=permutation,
             keep_in_memory=keep_in_memory,
-            load_from_cache_file=load_from_cache_file,
-            cache_file_name=cache_file_name,
+            indices_cache_file_name=indices_cache_file_name,
             writer_batch_size=writer_batch_size,
             verbose=verbose,
         )
 
-    def export(
```
Comment (thomwolf) [inline src/nlp/arrow_dataset.py:1334]: Moved this method without modification to keep all the samples re-ordering/selection methods (`select`, `sort`, `shuffle`, `shard`, `train_test_split`) in the same part of the file. Sorry for that.

File: src/nlp/arrow_dataset.py [hunk 2]
```diff
@@ -1419,8 +1482,8 @@ def train_test_split(
         generator: Optional[np.random.Generator] = None,
         keep_in_memory: bool = False,
         load_from_cache_file: bool = True,
-        train_cache_file_name: Optional[str] = None,
-        test_cache_file_name: Optional[str] = None,
+        train_indices_cache_file_name: Optional[str] = None,
```
Comment (thomwolf) [inline src/nlp/arrow_dataset.py:1485]: This is a bit long but I think it's important that the user does not mistake this cache for the dataset table cache.

**Review by lhoestq — COMMENTED** [review]
Very cool !
Things will be so much faster :)

a few comments:

File: src/nlp/arrow_dataset.py [hunk 3]
```diff
@@ -1080,81 +1178,128 @@ def map_function(batch, *args):
...
         if len(self.list_indexes()) > 0:
             raise DatasetTransformationNotAllowedError(
                 "Using `.select` on a dataset with attached indexes is not allowed. You can first run `.drop_index() to remove your index and then re-add it."
             )
+
         # If the array is empty we do nothing
         if len(self) == 0:
             return self
 
-        # Check if we've already cached this computation (indexed by a hash)
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply src/nlp/arrow_dataset.py:1119]: I feel like computing a hash on the indices and other stuff is actually more expensive than just replacing the indices with the new ones which kinda defeat the purpose of caching...

File: src/nlp/arrow_dataset.py [hunk 4]
```diff
@@ -1080,81 +1178,128 @@ def map_function(batch, *args):
...
-    def select(
+    def flatten_indices(
         self,
-        indices: Union[List[int], np.ndarray],
         keep_in_memory: bool = False,
-        load_from_cache_file: bool = True,
         cache_file_name: Optional[str] = None,
         writer_batch_size: Optional[int] = 1000,
-        reader_batch_size: Optional[int] = 1000,
+        features: Optional[Features] = None,
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply src/nlp/arrow_dataset.py:1231]: Free lunch :-)

File: src/nlp/arrow_dataset.py [hunk 5]
```diff
@@ -1221,22 +1365,22 @@ def sort(
...
-            if cache_file_name is None:
+            if indices_cache_file_name is None:
                 # we create a unique hash from the function, current dataset file and the mapping args
                 cache_kwargs = {
                     "column": column,
                     "reverse": reverse,
                     "kind": kind,
                     "keep_in_memory": keep_in_memory,
                     "load_from_cache_file": load_from_cache_file,
-                    "cache_file_name": cache_file_name,
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply src/nlp/arrow_dataset.py:1232]: Isn't the content of the dataset already handled by the hashes computed on `self._data_files` by `_get_cache_file_path`?

File: src/nlp/arrow_dataset.py [hunk 6]
```diff
@@ -1549,28 +1612,34 @@ def train_test_split(
                     "seed": seed,
                     "keep_in_memory": keep_in_memory,
                     "load_from_cache_file": load_from_cache_file,
-                    "train_cache_file_name": train_cache_file_name,
-                    "test_cache_file_name": test_cache_file_name,
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply src/nlp/arrow_dataset.py:1553]: I'm not sure about that because I feel like it's handled by the hashes on the indices and data files.
I add some tests on this.

[the remark this answers is not in the document]
Comment (lhoestq) [reply src/nlp/arrow_dataset.py:1553]: Yes you're right, it's taken into account in `previous_files_string`

File: src/nlp/arrow_dataset.py [hunk 7]
```diff
@@ -1301,21 +1444,21 @@ def shuffle(
...
         if self._data_files:
-            if cache_file_name is None:
+            if indices_cache_file_name is None:
                 # we create a unique hash from the function, current dataset file and the mapping args
                 cache_kwargs = {
                     "generator": generator,
                     "seed": seed,
                     "keep_in_memory": keep_in_memory,
                     "load_from_cache_file": load_from_cache_file,
-                    "cache_file_name": cache_file_name,
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/nlp/arrow_dataset.py:1311]: Yes you're right, it's taken into account in `previous_files_string`

File: src/nlp/arrow_dataset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/nlp/arrow_dataset.py:1232]: Yes you're right, it's taken into account in `previous_files_string`

**Comment by lhoestq** [comment]
Ok I fixed `concatenate_datasets` and added tests
Feel free to merge if it's good for you @thomwolf

**Comment by thomwolf** [comment]
Ok, adding some benchmarks for map/filters and then I'll merge

**Comment by thomwolf** [comment]
Warning from pytorch that we should maybe consider at some point @lhoestq:
```
/__w/nlp/nlp/src/nlp/arrow_dataset.py:648: UserWarning: The given NumPy array is not writeable,
and PyTorch does not support non-writeable tensors. This means you can write to the underlying
(supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to
protect its data or make it writeable before converting it to a tensor. This type of warning will be
suppressed for the rest of this program.
(Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
532
  return torch.tensor(x, **format_kwargs)
```

File: src/nlp/arrow_dataset.py [hunk 8]
```diff
@@ -717,7 +717,7 @@ def _map_indices(self, indices: Union[int, slice, pa.Array, Iterable]):
 
         # We can do a slice
         if array_indices is None:
-            return self._indices.column(0).slice(array_indices[0], array_indices[1] - array_indices[0])
+            return self._indices.column(0).slice(slice_indices[0], slice_indices[1] - slice_indices[0])
```
Comment (lhoestq) [inline src/nlp/arrow_dataset.py:720]: good catch !

**Comment by lhoestq** [comment]
> Warning from pytorch that we should maybe consider at some point @lhoestq:
> 
> ```
> /__w/nlp/nlp/src/nlp/arrow_dataset.py:648: UserWarning: The given NumPy array is not writeable,
> and PyTorch does not support non-writeable tensors. This means you can write to the underlying
> (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to
> protect its data or make it writeable before converting it to a tensor. This type of warning will be
> suppressed for the rest of this program.
> (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
> 532
>   return torch.tensor(x, **format_kwargs)
> ```

Not sure why we have that, it's probably linked o zero copy from arrow to numpy
